### PR TITLE
Model Couchbase Go library

### PR DESCRIPTION
**Pull request by owen-mc**
To do: ~add change note~.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,6 +3,11 @@ module github.com/github/codeql-go
 go 1.14
 
 require (
+	github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d
```
Comment (sauyon) [inline go.mod:6]: You should probably revert these changes.

File: ql/src/semmle/go/frameworks/NoSQL.qll [hunk 2]
```diff
@@ -104,6 +104,44 @@ module NoSQL {
         )
       }
     }
+
+    /**
+     * A query used in an API function acting on a `Bucket` or `Cluster` struct of v1 of
+     * the official Couchbase Go library, gocb.
+     */
+    private class CouchbaseV1Query extends Range {
```
Comment (sauyon) [inline ql/src/semmle/go/frameworks/NoSQL.qll:112]: I think I would prefer these to be together with the rest of the couchbase models.

File: ql/test/library-tests/semmle/go/frameworks/NoSQL/go.mod [hunk 3]
```diff
@@ -3,3 +3,15 @@ module main
 go 1.14
 
 require go.mongodb.org/mongo-driver v1.3.2
+
+require (
+	github.com/couchbase/gocb/v2 v2.2.0
+	github.com/github/depstubber v0.0.0-20201214172518-12c3da4b7c9d // indirect
```
Comment (sauyon) [inline ql/test/library-tests/semmle/go/frameworks/NoSQL/go.mod:9]: Hm, I'm surprised these are still showing up. Maybe you have to update depstubber manually with `go get`?

**Review by sauyon — COMMENTED** [review]
I'd forgotten that I'd meant to review this before I headed off :P

LGTM, with a few minor tidyness notes.

File: change-notes/2021-01-12-model-couchbase.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+lgtm,codescanning
+* Added support for [the offical Couchbase Go SDK library](https://github.com/couchbase/gocb), v1 and v2.
```
Comment (smowton) [inline change-notes/2021-01-12-model-couchbase.md:2]:
```suggestion
* Added support for [the offical Couchbase Go SDK library](https://github.com/couchbase/gocb), v1 and v2. The `go/sql-injection` query (which also handles non-SQL databases such as Couchbase) will now identify queries built from untrusted external input.
```

File: ql/src/semmle/go/frameworks/Couchbase.qll [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  /** Gets a module path for the official Couchbase Go SDK library. */
+  private string modulePath() {
+    result in [
+        "gopkg.in/couchbase/gocb", "github.com/couchbase/gocb", "github.com/couchbaselabs/gocb"
+      ]
+  }
+
+  private class QueryMethodV1 extends TaintTracking::FunctionModel, Method {
+    QueryMethodV1() {
+      exists(string queryName, string methodName |
```
Comment (smowton) [inline ql/src/semmle/go/frameworks/Couchbase.qll:29]:
```suggestion
      exists(string queryTypeName, string methodName |
```

File: vendor/modules.txt [hunk 6]
```diff
@@ -1,22 +1,21 @@
+# github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d
```
Comment (smowton) [inline vendor/modules.txt:1]: Looks like an accidental change?

File: ql/test/library-tests/semmle/go/frameworks/NoSQL/vendor/modules.txt [hunk 7]
```diff
@@ -1,3 +1,30 @@
 # go.mongodb.org/mongo-driver v1.3.2
 ## explicit
 go.mongodb.org/mongo-driver
+# github.com/couchbase/gocb/v2 v2.2.0
+## explicit
+github.com/couchbase/gocb/v2
+# github.com/github/depstubber v0.0.0-20201214172518-12c3da4b7c9d
```
Comment (smowton) [inline ql/test/library-tests/semmle/go/frameworks/NoSQL/vendor/modules.txt:7]: Remove depstubber deps everywhere

File: ql/test/library-tests/semmle/go/frameworks/CouchbaseV1/test.ql [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import go
+import TestUtilities.InlineExpectationsTest
+import semmle.go.security.SqlInjection
+
+class IsBad extends DataFlow::BarrierGuard, DataFlow::CallNode {
+  IsBad() { this.getTarget().getName() = "isBad" }
```
Comment (smowton) [inline ql/test/library-tests/semmle/go/frameworks/CouchbaseV1/test.ql:6]: Is this used anywhere?

File: ql/src/semmle/go/frameworks/Couchbase.qll [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        queryName = "AnalyticsQuery" and
+        methodName in [
+            "ContextId", "Deferred", "Pretty", "Priority", "RawParam", "ServerSideTimeout"
```
Comment (smowton) [inline ql/src/semmle/go/frameworks/Couchbase.qll:32]: Worth a quick comment here that these are present to support a fluent interface (e.g. `x.ContextId(xyz).Pretty(abc).Priority(def)...`), not because they're taint-relevant particularly

**Review by smowton — COMMENTED** [review]
Mostly looking good except the top-level vendor change.

Evaluation: let's do an LGTM run, pasting in the SQL query but hacked so that only Couchbase sinks are considered

File: go.mod (hunk 1 above)
Comment (owen-mc) [reply go.mod]: Oops 😬

File: ql/src/semmle/go/frameworks/NoSQL.qll (hunk 2 above)
Comment (owen-mc) [reply ql/src/semmle/go/frameworks/NoSQL.qll]: Sure

File: ql/test/library-tests/semmle/go/frameworks/NoSQL/go.mod (hunk 3 above)
Comment (owen-mc) [reply ql/test/library-tests/semmle/go/frameworks/NoSQL/go.mod]: It looks like it was stuck on the old `master` branch. But updating that and manually rebuilding didn't solve the problem.

Comment (smowton) [reply ql/test/library-tests/semmle/go/frameworks/NoSQL/go.mod]: I've found these always appear, and have just removed them by hand each time

Comment (gagliardetto) [reply ql/test/library-tests/semmle/go/frameworks/NoSQL/go.mod]: They always are added to go.mod because depstubber becomes a dependency to the package.

Depstubber becomes a dependency because stub.go is built inside the package folder (using reflection).

File: ql/test/library-tests/semmle/go/frameworks/CouchbaseV1/test.ql (hunk 8 above)
Comment (owen-mc) [reply ql/test/library-tests/semmle/go/frameworks/CouchbaseV1/test.ql:6]: Oops, copy-paste error

**Review by smowton — APPROVED** [review]

**Comment by owen-mc** [comment]
I've added a commit which removes the distinction between the paths for v1 and v2 of this package. It isn't needed, as the names are disjoint.
